notebooks/ConstraintSatisfactionProblems.ipynb

```diff
@@ -0,0 +1,514 @@
WMaPE84XV1c/s/efYc1dbUBAD8hAcKSQBL2UFlhKcOFgIoTZVWrta3WVe3wqyJqh11KrbXrq2KH/ax1a2tVlKWgiCgoiIAbQYYCYSdIwIQAF/L9cfSahiEjkEDe39Onz7k3N+ceFOHNuee8r6uTc3RUFF6HoJq5utqTf0SLEDIzNbG3s8VtsViccjWtP+4Cus90QCIGPp//7sqVebl5+//cV5CfT56vr68/sH8/QRAtLS1rVq/Ge0s3fPThkydPxo8Z894778x9dZ6+vj7R2nr/3j2E0KNHjz7e+GFCfPzWLeHvrV49ZuwYhND4CRMMDA2WvrVk186I0pKSA/v3I4R+2bXrcnIyvsu7779HUVN7c+HCrVvClyxbZmll9cO332WkXz8Xd/Z0ZCRC6OedEUVFRefizsbFxCKE/rf796KiojFjxvhOmnT6VOS8kFfKuGXv/Wc1l8vdsC5MSZaAGzKVaI+qj6/vxo8+amxsfH/VO09qn8i8uuWLL19dMN9/9uzZAXNWvftOdFTU2dg4hNDcV+cZGhrevXsHIUSlUt9b/T5CiPz2yLmfs/KdVQP7dfSeOk395RcBAAYbzzFjcIYv8gzO8EWGjN3UShCHDx6a4uNLxrIsNuunnTtzC/LXhIY6dGPRXYe0tbWDXwmJiz8XF38uMDiIPL8+dF1IYFB1dbV0rq4TpyNVJ7mBDEo//eaWSCRRMWfFYjE+9PRwgwW1imU7fERyasrAbOfsJrFYLKirw/NzLS0tZGWvVoKQIFRVVcVisaTTZbS0tAifPmUYdLpwE0+s4re0trZKJBJa9z6nCgQC/edrY5qampQkR0fY2lC/aVMVvrS/ubn5zYWvnzwdiQ/Xh66Ljoqa4OV18PChHf/9abzXBN9Jk7ilpX6TJr/+xhs0dRpCqLqqOiPj+htvLlq/cQNC6Ifvvjt04GDajQxdXd2c+/ffXPj6pCmTd/3yC0Jo1Yq3/7f3DzW1fvloLRfFJaVpzxNsz58X0s3vKADAYNQoEv34ww9krgCEEJvN/mX3b93JGFBbWxu2dq10osyfdu6cExgg9x8a3FLuu6tW5XW0Nvf4yRMqldxARn/9IqFQKDOm+ZGHWdm36p4ntAcAo9Pp5ONm6UK1VBqNRqOZm5vLRJbq6updhLMIIU1NTfItVCq1+z9H9KWWeitJOKu0vvnuW2cXl/S0tG+2bSNPPsx7KJFIPv500+bw8M3h4b/+vvtGdjYOZxFCC19/XSwWR505gxC6mnr13fffSzx/oa6uLvNGppu7uzKHswih5uYXqw4gnAVgaMMZvpJTU8hFCDjDV9jaUC6X28UbuVxuwCx/MpwNDA66npUZ/Eq/fAa2sLSIiz/3086dMuc3h4ercjiL+i+iRQjp6GhLJ/NKvpyqJE9yVVZ2VpaihwAGPTqdvnvP/wwNDQ/uP5AQH//sLIWCpNYSyLCytvb28T525EhZWdmwYcMWLFzY1tYWExV9/O+/Fry2oMO3KA8e71mxPYYiSkkDAAZe+wxfMdHROMNXh4sQsjIzp/j4ks/9f9q5c+euXUwms18HGfxKyPWsTOmVDNlZWUO1Clo39e/siLWVpbWVJW6LxeKbt+706+1AFxRefQr0VHqaUixAFwmFLS0t0mfMzMx+2f0blUZ99OgRPuPAcaBQKPv/3Ede09ra+uN335OHbyxalJeb99knnwQEBbJYrBmzZh46cKClucXI2HhgvopeE9TX44b+wJYpAQAoFs7wFRr2Ii16xI4dPl4To6OipC/Lyswkc2ax2KzklJTgVwZoqRiTyYyKjVmybBk+jImOXrFsmSoHtf3+vG/C+LFkXtKH+QXlFZX9fUfQIYVXnwI9JVN4VlEePnxYUV4uU3hi3PjxX3z5JXlobm4+y98/Ljb2i08/y83NffDgwYfrN3h5v8h5Mn36dCMjIzNzc5zS9Y0333z06NGrC+YP2FfRa3V1z5ZLsVj9O+MCAFA2OMNXRnZWUHAwPoMzfAX6z8aLEGTC2bj4+JcmMZAvGo325ZbNoWHr8OHV1KvS68FUTb9HtBQKZcpkH/Iw40YWJPMCoJsGuHxxe9FRUd99s722tva9Ve/czM6WfmnxkiULFi4kD7/57lsfX5+/jh0L9J+9/K0l/nNme/u8iGipNNr81157c9EifOg1caKPr4+3jw9SbkLhi/Q9MimWAQAqgszwRT7qzM3NneLj+5/33v/Pe+/jMw4cTtLly/290qAza0JDyaD24P4DP0fsUsgwFK6/ch3IuJ/z4O69HNy2t7P1cB89ADcF0sLWho60sVHBIiKDlxKmp5DR0tIiFov19F7Uhq2tra2rq7O2tpauvIDJJJEQi8XSVeWUE4/HT0xKxu2ggNk6Oh3U6QEAqIhWgoiLi8PJtqXh2VlFhbOknyMiInY82y6mmkkPBmiXsZMjR3rtAeQ9UIii5xX2AJALdXV16XAWIWRoaDhy5Mj24Sxql0RC+cNZhNATqRIe2tpaChwJAEDhqDRacEjI3Zz75CIEpDThLEJoTWgouab2g/dX41pCKmWAIlqZtQfp129A3oMBNtIG8gED0DPSiQ4oFIpiBwMAUAZa2tohc18hD7/97ntlCGexTz//DGc/qKmpCVurco9kBy4TJENfn8x7UFcnKCntKrUbkDvr4daKHgLoga5zH4KBUVxSihtGkCoEAIAQQqiVIDZ99DFuL1m2bMpUv66vH0g0Gu3Q0Wdl5K+mXpVJyzDkDWhu87FjPMhHjWnpGbBFbAiTSCRCoajD/4pLSvF/PB5f+jx8PwClIr0tzMRE2bOMAQAGxm+//oYT0bDYrE8//0zRw5HFZDLJ4gvbv97WKBJ1ff1QMqAlcGg0mrvbKLKk5J2792GL2ECKiY7esStCLl0RBNHU1Cyor29paWlubsYPZwX19WSqo76wtrIcNkzPwMBAf9gwbW0tVX7ay1TV8tzKgMfnk22mYVfF6gAAKqJRJIrYsQO3f929WznrCM4JDPjf77/n5
```

Comment:
These can be included as an Enum later instead of defining them separately in a class.